In [35]:
import pandas as pd
import numpy as np
import plotly.express as px
from numpy import isin
import pickle

In [36]:
data = pd.read_csv('data.csv')

In [37]:
data['marketcap'] = (data['PRC'] * data['SHROUT'])
data['PRC'] = data['PRC'].apply(lambda x : abs(x))
data['date1'] = data['date'].apply(lambda x : pd.Timestamp(str(x)))
data['RET1'] = data['RET'].replace('C',0).astype(float)

In [38]:
df = data
df

PERMNO      date  SHRCD  EXCHCD    NCUSIP TICKER  \
0         10000  19860131     10       3  68391610  OMFGA   
1         10000  19860228     10       3  68391610  OMFGA   
2         10000  19860331     10       3  68391610  OMFGA   
3         10000  19860430     10       3  68391610  OMFGA   
4         10000  19860530     10       3  68391610  OMFGA   
...         ...       ...    ...     ...       ...    ...   
2393463   93436  20211130     11       3  88160R10   TSLA   
2393464   93436  20211231     11       3  88160R10   TSLA   
2393465   93436  20220131     11       3  88160R10   TSLA   
2393466   93436  20220228     11       3  88160R10   TSLA   
2393467   93436  20220331     11       3  88160R10   TSLA   

                            COMNAM  PERMCO     CUSIP         PRC        RET  \
0        OPTIMUM MANUFACTURING INC    7952  68391610     4.37500          C   
1        OPTIMUM MANUFACTURING INC    7952  68391610     3.25000  -0.257143   
2        OPTIMUM MANUFACTURING INC    7952  68391610     4.43750   0.365385   
3        OPTIMUM MANUFACTURING INC    7952  68391610     4.00000  -0.098592   
4        OPTIMUM MANUFACTURING INC    7952  68391610     3.10938  -0.222656   
...                            ...     ...       ...         ...        ...   
2393463                  TESLA INC   53453  88160R10  1144.76001   0.027612   
2393464                  TESLA INC   53453  88160R10  1056.78003  -0.076855   
2393465                  TESLA INC   53453  88160R10   936.71997  -0.113609   
2393466                  TESLA INC   53453  88160R10   870.42999  -0.070768   
2393467                  TESLA INC   53453  88160R10  1077.59998   0.238009   

            SHROUT       RETX     marketcap      date1      RET1  
0           3680.0          C -1.610000e+04 1986-01-31  0.000000  
1           3680.0  -0.257143 -1.196000e+04 1986-02-28 -0.257143  
2           3680.0   0.365385 -1.633000e+04 1986-03-31  0.365385  
3           3793.0  -0.098592 -1.517200e+04 1986-04-30 -0.098592  
4           3793.0  -0.222656 -1.179388e+04 1986-05-30 -0.222656  
...            ...        ...           ...        ...       ...  
2393463  1004265.0   0.027612  1.149642e+09 2021-11-30  0.027612  
2393464  1033534.0  -0.076855  1.092218e+09 2021-12-31 -0.076855  
2393465  1033534.0  -0.113609  9.681319e+08 2022-01-31 -0.113609  
2393466  1033534.0  -0.070768  8.996190e+08 2022-02-28 -0.070768  
2393467  1033534.0   0.238009  1.113736e+09 2022-03-31  0.238009  

[2393468 rows x 16 columns]

# Removing negative market capitalizations

#(data['PRC'] < 0).value_counts(normalize= True)

lst = pd.DataFrame(data.groupby('COMNAM').nth(0)['marketcap'] < 0)
out = list(lst[lst.marketcap].index)
out

data[data['COMNAM'].isin(before)][data.date == 20090930].COMNAM

df = data[~data['COMNAM'].isin(out)] #remove negative marketcaps
df

# Remove microcap stocks

In [39]:
from numpy import isin


microcaps = pd.DataFrame(df.groupby('TICKER').nth(0))
microcaps = list(microcaps[microcaps.marketcap < microcaps[microcaps['EXCHCD']==1].marketcap.quantile(0.2)].index)
#px.histogram(df.groupby('COMNAM').nth(0), 'marketcap', nbins = 50)


In [40]:
df = df[~df['TICKER'].isin(microcaps)]
df

PERMNO      date  SHRCD  EXCHCD    NCUSIP TICKER  \
911       10006  19840131     10       1  00080010    ACF   
912       10006  19840229     10       1  00080010    ACF   
913       10006  19840330     10       1  00080010    ACF   
914       10006  19840430     10       1  00080010    ACF   
915       10006  19840531     10       1  00080010    ACF   
...         ...       ...    ...     ...       ...    ...   
2393463   93436  20211130     11       3  88160R10   TSLA   
2393464   93436  20211231     11       3  88160R10   TSLA   
2393465   93436  20220131     11       3  88160R10   TSLA   
2393466   93436  20220228     11       3  88160R10   TSLA   
2393467   93436  20220331     11       3  88160R10   TSLA   

                       COMNAM  PERMCO     CUSIP         PRC        RET  \
911      A C F INDUSTRIES INC   22156  00080010    45.62500  -0.087500   
912      A C F INDUSTRIES INC   22156  00080010    47.75000   0.054247   
913      A C F INDUSTRIES INC   22156  00080010    51.00000   0.068063   
914      A C F INDUSTRIES INC   22156  00080010    52.62500   0.031863   
915      A C F INDUSTRIES INC   22156  00080010    52.62500   0.000000   
...                       ...     ...       ...         ...        ...   
2393463             TESLA INC   53453  88160R10  1144.76001   0.027612   
2393464             TESLA INC   53453  88160R10  1056.78003  -0.076855   
2393465             TESLA INC   53453  88160R10   936.71997  -0.113609   
2393466             TESLA INC   53453  88160R10   870.42999  -0.070768   
2393467             TESLA INC   53453  88160R10  1077.59998   0.238009   

            SHROUT       RETX     marketcap      date1      RET1  
911         8385.0  -0.087500  3.825656e+05 1984-01-31 -0.087500  
912         8385.0   0.046575  4.003838e+05 1984-02-29  0.054247  
913         8385.0   0.068063  4.276350e+05 1984-03-30  0.068063  
914         8385.0   0.031863  4.412606e+05 1984-04-30  0.031863  
915         8385.0   0.000000  4.412606e+05 1984-05-31  0.000000  
...            ...        ...           ...        ...       ...  
2393463  1004265.0   0.027612  1.149642e+09 2021-11-30  0.027612  
2393464  1033534.0  -0.076855  1.092218e+09 2021-12-31 -0.076855  
2393465  1033534.0  -0.113609  9.681319e+08 2022-01-31 -0.113609  
2393466  1033534.0  -0.070768  8.996190e+08 2022-02-28 -0.070768  
2393467  1033534.0   0.238009  1.113736e+09 2022-03-31  0.238009  

[867841 rows x 16 columns]

# Remove penny stocks

penny = pd.DataFrame(df.groupby('COMNAM').nth(0)['PRC'] < 5)
penny = pd.DataFrame(df.groupby('COMNAM').agg({'PRC' : 'min'}) < 5)
pennystocks = list(penny[penny.PRC].index)

df = df[~df['COMNAM'].isin(pennystocks)]

In [41]:
df = df[df['PRC'] >= 5]
df

PERMNO      date  SHRCD  EXCHCD    NCUSIP TICKER  \
911       10006  19840131     10       1  00080010    ACF   
912       10006  19840229     10       1  00080010    ACF   
913       10006  19840330     10       1  00080010    ACF   
914       10006  19840430     10       1  00080010    ACF   
915       10006  19840531     10       1  00080010    ACF   
...         ...       ...    ...     ...       ...    ...   
2393463   93436  20211130     11       3  88160R10   TSLA   
2393464   93436  20211231     11       3  88160R10   TSLA   
2393465   93436  20220131     11       3  88160R10   TSLA   
2393466   93436  20220228     11       3  88160R10   TSLA   
2393467   93436  20220331     11       3  88160R10   TSLA   

                       COMNAM  PERMCO     CUSIP         PRC        RET  \
911      A C F INDUSTRIES INC   22156  00080010    45.62500  -0.087500   
912      A C F INDUSTRIES INC   22156  00080010    47.75000   0.054247   
913      A C F INDUSTRIES INC   22156  00080010    51.00000   0.068063   
914      A C F INDUSTRIES INC   22156  00080010    52.62500   0.031863   
915      A C F INDUSTRIES INC   22156  00080010    52.62500   0.000000   
...                       ...     ...       ...         ...        ...   
2393463             TESLA INC   53453  88160R10  1144.76001   0.027612   
2393464             TESLA INC   53453  88160R10  1056.78003  -0.076855   
2393465             TESLA INC   53453  88160R10   936.71997  -0.113609   
2393466             TESLA INC   53453  88160R10   870.42999  -0.070768   
2393467             TESLA INC   53453  88160R10  1077.59998   0.238009   

            SHROUT       RETX     marketcap      date1      RET1  
911         8385.0  -0.087500  3.825656e+05 1984-01-31 -0.087500  
912         8385.0   0.046575  4.003838e+05 1984-02-29  0.054247  
913         8385.0   0.068063  4.276350e+05 1984-03-30  0.068063  
914         8385.0   0.031863  4.412606e+05 1984-04-30  0.031863  
915         8385.0   0.000000  4.412606e+05 1984-05-31  0.000000  
...            ...        ...           ...        ...       ...  
2393463  1004265.0   0.027612  1.149642e+09 2021-11-30  0.027612  
2393464  1033534.0  -0.076855  1.092218e+09 2021-12-31 -0.076855  
2393465  1033534.0  -0.113609  9.681319e+08 2022-01-31 -0.113609  
2393466  1033534.0  -0.070768  8.996190e+08 2022-02-28 -0.070768  
2393467  1033534.0   0.238009  1.113736e+09 2022-03-31  0.238009  

[742656 rows x 16 columns]

# Let's now pivot the data to plot the time series

In [42]:
ts = pd.DataFrame(df.pivot_table(
    values= 'RET1',
index ='TICKER', 
columns= ['date1']
))

In [44]:
ts

date1   1984-01-31  1984-02-29  1984-03-30  1984-04-30  1984-05-31  \
TICKER                                                               
AA       -0.062953   -0.101796    0.080000   -0.125309   -0.035587   
AAC      -0.132075   -0.130435         NaN         NaN         NaN   
AACC           NaN         NaN         NaN         NaN         NaN   
AADI           NaN         NaN         NaN         NaN         NaN   
AAI            NaN         NaN         NaN         NaN         NaN   
...            ...         ...         ...         ...         ...   
ZWS            NaN         NaN         NaN         NaN         NaN   
ZY       -0.190751    0.016429    0.035211    0.000000    0.013197   
ZYNE           NaN         NaN         NaN         NaN         NaN   
ZYXI           NaN         NaN         NaN         NaN         NaN   
ZZ             NaN         NaN         NaN         NaN         NaN   

date1   1984-06-29  1984-07-31  1984-08-31  1984-09-28  1984-10-31  ...  \
TICKER                                                              ...   
AA        0.003690   -0.068382    0.139442   -0.076923    0.107576  ...   
AAC            NaN         NaN         NaN         NaN         NaN  ...   
AACC           NaN         NaN         NaN         NaN         NaN  ...   
AADI           NaN         NaN         NaN         NaN         NaN  ...   
AAI            NaN         NaN         NaN         NaN         NaN  ...   
...            ...         ...         ...         ...         ...  ...   
ZWS            NaN         NaN         NaN         NaN         NaN  ...   
ZY        0.111111    0.033333    0.228387   -0.063158    0.025281  ...   
ZYNE           NaN         NaN         NaN         NaN         NaN  ...   
ZYXI           NaN         NaN         NaN         NaN         NaN  ...   
ZZ             NaN         NaN         NaN         NaN         NaN  ...   

date1   2021-06-30  2021-07-30  2021-08-31  2021-09-30  2021-10-29  \
TICKER                                                               
AA       -0.071338    0.089848    0.105106    0.102998   -0.059052   
AAC            NaN         NaN         NaN         NaN         NaN   
AACC           NaN         NaN         NaN         NaN         NaN   
AADI           NaN         NaN   -0.140263    0.127447   -0.081035   
AAI            NaN         NaN         NaN         NaN         NaN   
...            ...         ...         ...         ...         ...   
ZWS            NaN         NaN         NaN         NaN    0.097639   
ZY        0.235639   -0.130467   -0.622880    0.003811   -0.195140   
ZYNE     -0.068662         NaN         NaN         NaN         NaN   
ZYXI      0.020368   -0.105602   -0.034557   -0.150634    0.100088   
ZZ             NaN         NaN         NaN         NaN         NaN   

date1   2021-11-30  2021-12-31  2022-01-31  2022-02-28  2022-03-31  
TICKER                                                              
AA        0.012622    0.280464   -0.048171    0.328513    0.196310  
AAC            NaN         NaN         NaN         NaN         NaN  
AACC           NaN         NaN         NaN         NaN         NaN  
AADI     -0.203779    0.123778   -0.213251   -0.083158   -0.025832  
AAI            NaN         NaN         NaN         NaN         NaN  
...            ...         ...         ...         ...         ...  
ZWS      -0.033076    0.038516   -0.160989    0.065815    0.088561  
ZY       -0.145283   -0.261589   -0.222721         NaN         NaN  
ZYNE           NaN         NaN         NaN         NaN         NaN  
ZYXI      0.021548   -0.221094   -0.116148   -0.205808   -0.009539  
ZZ             NaN         NaN         NaN         NaN         NaN  

[7187 rows x 459 columns]

# Output after filtering 

After filtering for penny stocks and microcaps, there are 7187 stocks remaining for our portfolio, out of the initial 27 840 listed on the stock exchanges we are considering.

In [46]:
px.line(ts.iloc[:75].T, color = 'TICKER', labels = dict(value = 'Return', date1 = 'Date', COMNAM = 'Stock name'), title = 'Return of a sample of stocks from the filtered dataset')

In [47]:
ts

date1   1984-01-31  1984-02-29  1984-03-30  1984-04-30  1984-05-31  \
TICKER                                                               
AA       -0.062953   -0.101796    0.080000   -0.125309   -0.035587   
AAC      -0.132075   -0.130435         NaN         NaN         NaN   
AACC           NaN         NaN         NaN         NaN         NaN   
AADI           NaN         NaN         NaN         NaN         NaN   
AAI            NaN         NaN         NaN         NaN         NaN   
...            ...         ...         ...         ...         ...   
ZWS            NaN         NaN         NaN         NaN         NaN   
ZY       -0.190751    0.016429    0.035211    0.000000    0.013197   
ZYNE           NaN         NaN         NaN         NaN         NaN   
ZYXI           NaN         NaN         NaN         NaN         NaN   
ZZ             NaN         NaN         NaN         NaN         NaN   

date1   1984-06-29  1984-07-31  1984-08-31  1984-09-28  1984-10-31  ...  \
TICKER                                                              ...   
AA        0.003690   -0.068382    0.139442   -0.076923    0.107576  ...   
AAC            NaN         NaN         NaN         NaN         NaN  ...   
AACC           NaN         NaN         NaN         NaN         NaN  ...   
AADI           NaN         NaN         NaN         NaN         NaN  ...   
AAI            NaN         NaN         NaN         NaN         NaN  ...   
...            ...         ...         ...         ...         ...  ...   
ZWS            NaN         NaN         NaN         NaN         NaN  ...   
ZY        0.111111    0.033333    0.228387   -0.063158    0.025281  ...   
ZYNE           NaN         NaN         NaN         NaN         NaN  ...   
ZYXI           NaN         NaN         NaN         NaN         NaN  ...   
ZZ             NaN         NaN         NaN         NaN         NaN  ...   

date1   2021-06-30  2021-07-30  2021-08-31  2021-09-30  2021-10-29  \
TICKER                                                               
AA       -0.071338    0.089848    0.105106    0.102998   -0.059052   
AAC            NaN         NaN         NaN         NaN         NaN   
AACC           NaN         NaN         NaN         NaN         NaN   
AADI           NaN         NaN   -0.140263    0.127447   -0.081035   
AAI            NaN         NaN         NaN         NaN         NaN   
...            ...         ...         ...         ...         ...   
ZWS            NaN         NaN         NaN         NaN    0.097639   
ZY        0.235639   -0.130467   -0.622880    0.003811   -0.195140   
ZYNE     -0.068662         NaN         NaN         NaN         NaN   
ZYXI      0.020368   -0.105602   -0.034557   -0.150634    0.100088   
ZZ             NaN         NaN         NaN         NaN         NaN   

date1   2021-11-30  2021-12-31  2022-01-31  2022-02-28  2022-03-31  
TICKER                                                              
AA        0.012622    0.280464   -0.048171    0.328513    0.196310  
AAC            NaN         NaN         NaN         NaN         NaN  
AACC           NaN         NaN         NaN         NaN         NaN  
AADI     -0.203779    0.123778   -0.213251   -0.083158   -0.025832  
AAI            NaN         NaN         NaN         NaN         NaN  
...            ...         ...         ...         ...         ...  
ZWS      -0.033076    0.038516   -0.160989    0.065815    0.088561  
ZY       -0.145283   -0.261589   -0.222721         NaN         NaN  
ZYNE           NaN         NaN         NaN         NaN         NaN  
ZYXI      0.021548   -0.221094   -0.116148   -0.205808   -0.009539  
ZZ             NaN         NaN         NaN         NaN         NaN  

[7187 rows x 459 columns]

# Store the list of tickers so we can match it with other datasets

In [53]:
l = list(df['TICKER'].unique())
filename = 'stocks.pkl'
with open(filename, 'wb') as f : 
    pickle.dump(l, f)